src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs

```diff
@@ -51,200 +64,177 @@ public Task WarmUpCacheAsync(Project? project, CancellationToken cancellationTok
+                {
+                    isPartialResult = true;
                 }
             }
 
+            var originCompilation = await currentProject.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var editorBrowsableInfo = new Lazy<EditorBrowsableInfo>(() => new EditorBrowsableInfo(originCompilation));
             foreach (var peReference in currentProject.MetadataReferences.OfType<PortableExecutableReference>())
             {
-                if (HasGlobalAlias(peReference) &&
-                    currentCompilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol assembly &&
-                    TryGetCacheForPEReference(solution, assembly, editorBrowsableInfo, peReference, forceCacheCreation, cancellationToken, out cacheResult))
+                // Can't cache items for reference with null key. We don't want risk potential perf regression by 
+                // making those items repeatedly, so simply not returning anything from this assembly, until 
+                // we have a better understanding on this scenario.
+                var peReferenceKey = GetPEReferenceCacheKey(peReference);
+                if (peReferenceKey is null || !HasGlobalAlias(peReference.Properties.Aliases))
+                    continue;
+
+                if (forceCacheCreation)
                 {
-                    if (cacheResult.HasValue)
+                    if (TryGetUpToDateCacheForPEReference(originCompilation, solution, editorBrowsableInfo.Value, peReference, cancellationToken, out var upToDateCacheEntry))
                     {
-                        builder.Add(cacheResult.Value);
-                    }
-                    else
-                    {
-                        isPartialResult = true;
+                        resultBuilder.Add(upToDateCacheEntry);
                     }
                 }
+                else if (CacheService.PEItemsCache.TryGetValue(peReferenceKey, out var cacheEntry))
+                {
+                    resultBuilder.Add(cacheEntry);
+                }
+                else
+                {
+                    isPartialResult = true;
+                }
             }
 
-            return (builder.ToImmutable(), isPartialResult);
+            if (!forceCacheCreation)
+                _workQueue.AddWork(currentProject.Id);
 
-            static bool HasGlobalAlias(MetadataReference? metadataReference)
-                => metadataReference != null && (metadataReference.Properties.Aliases.IsEmpty || metadataReference.Properties.Aliases.Any(alias => alias == MetadataReferenceProperties.GlobalAlias));
+            return (resultBuilder.ToImmutable(), isPartialResult);
         }
 
+        private async ValueTask UpdateCacheAsync(ImmutableArray<ProjectId> projectIds, CancellationToken cancellationToken)
+        {
+            foreach (var projectId in projectIds)
+            {
+                var project = _workspace.CurrentSolution.GetProject(projectId);
```

Comment:
pull out _workspace.CurrentSolution out of loop.  it makes me queezy that this could be changing :)

Reply:
Good catch, this is not intended :)